### PR TITLE
New Adapter: liftoff

### PR DESCRIPTION
**Pull request by Vungle-GordonTian**
https://github.com/prebid/prebid.github.io/pull/4757

### CODE REVIEW

File: adapters/liftoff/liftoff.go
Comment (Vungle-GordonTian) [reply adapters/liftoff/liftoff.go:41]: Yes, it's unused. But the interface must provide the parameters.

File: adapters/liftoff/liftoff.go
Comment (SyntaxNode) [inline adapters/liftoff/liftoff.go:36]: Please do not use this comment style. Go standards are to use two slashes followed by the function name and then the comment, such as:

```
// Builder builds a new instance of the Liftoff adapter for the given bidder with the given config.
```

File: adapters/liftoff/liftoff.go
Comment (SyntaxNode) [inline adapters/liftoff/liftoff.go:32]: What's vungle?

File: adapters/liftoff/liftoff.go
Comment (SyntaxNode) [inline adapters/liftoff/liftoff.go:53]: This check is not necessary. Prebid Server will not send your adapter requests for media types not defined in the yaml file. Please remove this check.

File: adapters/liftoff/liftoff.go
Comment (SyntaxNode) [inline adapters/liftoff/liftoff.go:83]: Can you move these length constraints to the bidder params json file? Then you don't need to worry about error handling here.

File: adapters/liftoff/liftoff.go
Comment (SyntaxNode) [inline adapters/liftoff/liftoff.go:88]: Please use errors.New instead of fmt.Errorf when there are no variables / macros in the error string.

**Review by SyntaxNode — COMMENTED** [review]
Great first pass. Thank you for properly handling all errors.

File: adapters/liftoff/liftoff.go
Comment (Vungle-GordonTian) [reply adapters/liftoff/liftoff.go:32]: `Vungle` is our company name before merging with liftoff. In our server code it's a field name in `imp.ext.vungle` which holds our external partners' token and other info.

File: adapters/liftoff/liftoff.go
Comment (Vungle-GordonTian) [reply adapters/liftoff/liftoff.go:53]: Got it.

File: adapters/liftoff/liftoff.go
Comment (Vungle-GordonTian) [reply adapters/liftoff/liftoff.go:83]: Got it.

Comment (Vungle-GordonTian) [reply adapters/liftoff/liftoff.go:83]: @SyntaxNode I tried moving it into yaml with `"minLength": 1`, but when I run the same UT test, got the different reuslt from before. It produce request for case `missing_appid_or_placementid.json`. Is it right?

**Comment by Vungle-GordonTian** [comment]
@SyntaxNode PTAL, I think I've fix your comments.

File: go.mod
Comment (Sonali-More-Xandr) [inline go.mod:33]: Can you please undo changes made to go.mod and go.sum?

Comment (Vungle-GordonTian) [reply go.mod:33]: @Sonali-More-Xandr Got it. It changes perhaps because I did the local end to end test with the command `go build .`

Comment (Vungle-GordonTian) [reply go.mod]: DONE

File: static/bidder-info/liftoff.yaml
Comment (gargcreation1992) [inline static/bidder-info/liftoff.yaml:3]: sent email to this email address for verification.

Comment (gargcreation1992) [reply static/bidder-info/liftoff.yaml:3]: @Vungle-GordonTian email is bounced. Could you please re-check if this email ID is working.

File: openrtb_ext/imp_liftoff.go
Comment (Sonali-More-Xandr) [inline openrtb_ext/imp_liftoff.go:6]: Please remove the `omitempty` tag as all the parameters are defined as required in the `bidder-params/liftoff.json`

Comment (Vungle-GordonTian) [reply openrtb_ext/imp_liftoff.go]: @Sonali-More-Xandr DONE

File: static/bidder-info/liftoff.yaml
Comment (Vungle-GordonTian) [reply static/bidder-info/liftoff.yaml:3]: @gargcreation1992 I'm sure this is the valid and correct email address. liftoff.io is blocked only in China, otherwise it should be ok. Can u provide your email info such as address and send server?

Comment (gargcreation1992) [reply static/bidder-info/liftoff.yaml:3]: Here is the reply that we have received - 
```
Hello [prebid@microsoft.com](mailto:prebid@microsoft.com),

We're writing to let you know that the group you tried to contact (platform-ssp) may not exist, or you may not have permission to post messages to the group. A few more details on why you weren't able to post:

 * You might have spelled or formatted the group name incorrectly.
 * The owner of the group may have removed this group.
 * You may need to join the group before receiving permission to post.
 * This group may not be open to posting.
 ```

Comment (Vungle-GordonTian) [reply static/bidder-info/liftoff.yaml:3]: @gargcreation1992 Many thanks for the detailed info. I'll refer to our IT to find out the root cause.

Comment (Vungle-GordonTian) [reply static/bidder-info/liftoff.yaml:3]: @gargcreation1992 I've reach out to our IT and they have now adjusted the permissions for [platform-ssp@liftoff.io](mailto:platform-ssp@liftoff.io) to accept incoming emails from external senders.

Could you try sending an email again and it should all be good.

Comment (Vungle-GordonTian) [reply static/bidder-info/liftoff.yaml:3]: done. @gargcreation1992

File: adapters/liftoff/liftoff.go
Comment (gargcreation1992) [inline adapters/liftoff/liftoff.go:125]: this is correct but you might also wanna use built-in methods - 
```
if adapters.IsResponseStatusCodeNoContent(response) {
		return nil, nil
	}

	if err := adapters.CheckResponseStatusCodeForErrors(response); err != nil {
		return nil, []error{err}
	}
```

File: adapters/liftoff/liftoff.go
Comment (gargcreation1992) [inline adapters/liftoff/liftoff.go:137]: this is correct but you might wanna consider the following code to avoid creating a new temp variable - 
```
for i := range seatBid.Bid {
    b := &adapters.TypedBid{
				Bid:     &seatBid.Bid[i],
				BidType: openrtb_ext.BidTypeVideo,
			}
       // more code

}
```

Comment (Vungle-GordonTian) [reply adapters/liftoff/liftoff.go]: Done optimizing

**Review by gargcreation1992 — APPROVED** [review]

**Comment by Vungle-GordonTian** [comment]
@gargcreation1992 I'm wondering what's the next move since this PR and [user documentation PR](https://github.com/prebid/prebid.github.io/pull/4757) are both approved?

**Comment by gargcreation1992** [comment]
> @gargcreation1992 I'm wondering what's the next move since this PR and [user documentation PR](https://github.com/prebid/prebid.github.io/pull/4757) are both approved? And I've done local end to end test.

@Vungle-GordonTian thanks for being patient on this one. There are two required approvals before we merge this one in. Someone from the team will be taking a look at this PR soon.

**Review by Sonali-More-Xandr — APPROVED** [review]

**Comment by Vungle-GordonTian** [comment]
@gargcreation1992 Per as the [documentation](https://docs.prebid.org/prebid-server/developers/add-new-bidder-go.html) said, you will verify the adapter works correctly on a technical level during the code review, now since the code has been merged, I want to confirm if the technical level verification(I would reckon that as e2e test, right?) has been done.

**Comment by gargcreation1992** [comment]
> @gargcreation1992 Per as the [documentation](https://docs.prebid.org/prebid-server/developers/add-new-bidder-go.html) said, you will verify the adapter works correctly on a technical level during the code review, now since the code has been merged, I want to confirm if the technical level verification(I would reckon that as e2e test, right?) has been done.

Please refer to this part of documentation for e2e test - https://docs.prebid.org/prebid-server/developers/add-new-bidder-go.html#manual-end-to-end-tests

The general guideline is to run prebid-server in your local and test your adapter using a sample request as mentioned in the docs above.

**Comment by Vungle-GordonTian** [comment]
> > @gargcreation1992 Per as the [documentation](https://docs.prebid.org/prebid-server/developers/add-new-bidder-go.html) said, you will verify the adapter works correctly on a technical level during the code review, now since the code has been merged, I want to confirm if the technical level verification(I would reckon that as e2e test, right?) has been done.
> 
> Please refer to this part of documentation for e2e test - https://docs.prebid.org/prebid-server/developers/add-new-bidder-go.html#manual-end-to-end-tests
> 
> The general guideline is to run prebid-server in your local and test your adapter using a sample request as mentioned in the docs above.

Yeah, I've done the local e2e test already.

**Comment by Vungle-GordonTian** [comment]
> > > @gargcreation1992 Per as the [documentation](https://docs.prebid.org/prebid-server/developers/add-new-bidder-go.html) said, you will verify the adapter works correctly on a technical level during the code review, now since the code has been merged, I want to confirm if the technical level verification(I would reckon that as e2e test, right?) has been done.
> > 
> > 
> > Please refer to this part of documentation for e2e test - https://docs.prebid.org/prebid-server/developers/add-new-bidder-go.html#manual-end-to-end-tests
> > The general guideline is to run prebid-server in your local and test your adapter using a sample request as mentioned in the docs above.
> 
> @gargcreation1992 Yeah, I've done the local e2e test already.

@gargcreation1992 Hi, I am wondering when will our adapter be on production on Xandr side? Will we be informed of it through the group email?

**Pull request merged**
